clients/src/main/java/org/apache/kafka/clients/ClientTelemetryState.java

```diff
@@ -0,0 +1,167 @@
+            TERMINATING_PUSH_NEEDED, TERMINATED));
+
+        /*
+         A successful push should transition client to push needed which sends the next telemetry
+         metrics after the elapsed wait interval. But, if the push fails (network issues, the
+         subscription changed, etc.) then client should go back to subscription needed state and
+         request the next subscription.
+
+         However, it's still possible that client doesn't get very far before terminating.
+        */
+        VALID_NEXT_STATES.put(
+            PUSH_IN_PROGRESS, Arrays.asList(PUSH_NEEDED, SUBSCRIPTION_NEEDED, TERMINATING_PUSH_NEEDED,
+                TERMINATED));
+
+        /*
+         If client is moving out of this state, then try to send last metrics push.
+
+         However, it's still possible that client doesn't get very far before terminating.
+        */
+        VALID_NEXT_STATES.put(
+            TERMINATING_PUSH_NEEDED, Arrays.asList(TERMINATING_PUSH_IN_PROGRESS, TERMINATED));
+
+        /*
+         Client should only be transited to terminated state.
+        */
+        VALID_NEXT_STATES.put(TERMINATING_PUSH_IN_PROGRESS, Collections.singletonList(TERMINATED));
+
+        /*
+         Client should never be able to transition out of terminated state.
+        */
+        VALID_NEXT_STATES.put(TERMINATED, Collections.emptyList());
+    }
+
+    /**
+     * Validates that the <code>newState</code> is one of the valid transition from the current
+     * {@code TelemetryState}.
+     *
+     * @param newState State into which the telemetry client requesting to transition; must be
+     *                 non-<code>null</code>
+     * @return {@code TelemetryState} <code>newState</code> if validation succeeds. Returning
+     * <code>newState</code> helps state assignment chaining.
+     * @throws IllegalClientTelemetryStateException if the state transition validation fails.
+     */
+
+    public ClientTelemetryState validateTransition(ClientTelemetryState newState) {
+        List<ClientTelemetryState> allowableStates = VALID_NEXT_STATES.get(this);
+
+        if (allowableStates != null && allowableStates.contains(newState)) {
+            return newState;
+        }
+
+        // State transition validation failed, construct error message and throw exception.
+        String validStatesClause;
+        if (allowableStates != null && !allowableStates.isEmpty()) {
+            validStatesClause = String.format("the valid telemetry state transitions from %s are: %s",
+                this,
+                Utils.join(allowableStates, ", "));
+        } else {
+            validStatesClause = String.format("there are no valid telemetry state transitions from %s", this);
+        }
+
+        String message = String.format("Invalid telemetry state transition from %s to %s; %s",
+            this,
+            newState,
+            validStatesClause);
+
+        throw new IllegalClientTelemetryStateException(message);
```

Comment:
Can we just throw IllegalStateException? Example from KafkaStream
```
                throw new IllegalStateException("Stream-client " + clientId + ": Unexpected state transition from " + oldState + " to " + newState);
```

Reply:
Done.